core/cache/modelsummarywatcher_test.go

```diff
@@ -0,0 +1,578 @@
+		{
+			UUID:      "model-1-uuid",
+			Namespace: "test-admin",
+			Name:      "model-1",
+			Admins:    []string{"test-admin"},
+			Status:    cache.StatusGreen,
+			// We didn't actually remove the machine, or unit yet.
+			// Yes I know in theory this can't happen, but hey, this is a test.
+			MachineCount: 1,
+			UnitCount:    1,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestAddingUnitIsChange(c *gc.C) {
+	watcher := s.controller.WatchAllModels()
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the intial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.UnitChange{
+		ModelUUID: "model-1-uuid",
+		Name:      "magic/1",
+		Life:      life.Alive,
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:             "model-1-uuid",
+			Namespace:        "test-admin",
+			Name:             "model-1",
+			Admins:           []string{"test-admin"},
+			Status:           cache.StatusGreen,
+			MachineCount:     1,
+			ApplicationCount: 1,
+			UnitCount:        2,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestRemovingUnitIsChange(c *gc.C) {
+	watcher := s.controller.WatchAllModels()
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the intial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.RemoveUnit{
+		ModelUUID: "model-1-uuid",
+		Name:      "magic/0",
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:      "model-1-uuid",
+			Namespace: "test-admin",
+			Name:      "model-1",
+			Admins:    []string{"test-admin"},
+			Status:    cache.StatusGreen,
+			// We didn't actually remove the machine, or application yet.
+			MachineCount:     1,
+			ApplicationCount: 1,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestChangesToOneModelCoalesced(c *gc.C) {
+	watcher := s.controller.WatchAllModels()
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the intial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.RemoveUnit{
+		ModelUUID: "model-1-uuid",
+		Name:      "magic/0",
+	}, s.events)
+	s.ProcessChange(c, cache.RemoveApplication{
+		ModelUUID: "model-1-uuid",
+		Name:      "magic",
+	}, s.events)
+	s.ProcessChange(c, cache.RemoveMachine{
+		ModelUUID: "model-1-uuid",
+		Id:        "0",
+	}, s.events)
+	s.ProcessChange(c, cache.ApplicationChange{
+		ModelUUID: "model-2-uuid",
+		Name:      "foo",
+		Life:      life.Alive,
+	}, s.events)
+
+	// Make sure all the published summary events have been handled.
+	// We know that the events are consumed in the order that they are
+	// published, and that any publishing will take place during the
+	// above ProcessChange calls above, so we only need to wait on the
+	// last publish event from model 2.
+	cache.WaitForModelSummaryHandled(c, s.controller, "model-2-uuid")
+
+	update := s.next(c, changes, "model-2-uuid")
```

Comment:
Why is it ok to explicitly wait for 'model-2-uuid' above, and then wait again in next() here?

Reply:
It is ok because the channel has been closed. However this is a dupe, removing first wait.